### PR TITLE
03b modeling steps cleaner

### PR DESCRIPTION
**Pull request by galengorski**

### CODE REVIEW

**Comment by galengorski** [comment]
This is now ready for review, I have tested it in a newly built environment and there are a few known issues, but it should run ok. Since this PR got a little bit too big, there are  a lot of files. I would appreciate feedback on the `run_model.py` script as that is the newest and most important in the PR.

Steps for running the model:

1. clone this branch
2. add the file `953860.zip` into the 01_fetch/in folder, the file can be found on S3 in drb_estuary_salinity/01_fetch/in
3. from within the github cloned directory create the environment using `conda env create -f environment.yaml`
4. run `snakemake -s Snakefile_fetch_munge -j` (-j runs the job on the available number of cpus cores, use -j 2 for fewer)
5. you might have to rerun the same command if there is an error that pops up, this is because snakemake doesn't run rules in order and some directories need to be created
6. now open the file `03b_model/model_config.yaml` and change `n_epochs` to a small number say 5, and change the `run_id` to whatever you want to name the test run, say `Test_Run`
7. run `snakemake -s Snakefile_b_ml_model_baseline -j`, you should see the training progress in the command window and you should have model results written to `03b_model/out/Test_Run/`

File: 03b_model/src/run_model.py [hunk 1]
```diff
@@ -54,19 +101,58 @@ def select_inputs_targets(inputs, target, train_start_date, test_end_date, out_d
...
+            #and the antecedant discharge of the two combined as another predictor
+            inputs_df = inputs_df.drop(columns = 'discharge')
+        
+        inputs_df = inputs_df.dropna()
+    
     
     #read in the salt front record
     target_df = pd.read_csv(os.path.join('03a_it_analysis', 'in', 'saltfront.csv'), parse_dates = True, index_col = 'datetime')
-    target_df = target_df['saltfront_daily'].to_frame()
+    target_df = target_df['saltfront7_weekly'].to_frame()
```
Comment (jds485) [inline 03b_model/src/run_model.py:138]: Would it help to make the variable to pull ('saltfront7_weekly') a function argument? Something like target_variable. Comment applies to all other functions in this code that specify target names.

File: 03b_model/src/run_model.py [hunk 2]
```diff
@@ -157,50 +282,160 @@ def prep_input_target_data(inputs_xarray, target_xarray,
...
+                       test_start_date, test_end_date, inc_ante):
+    '''
+    Write model parameters to directory where model results will be saved
+
+    Parameters
+    ----------
+    out_dir : str
+        directory where model results will be saved
+    run_id : str
+        sub-directory within out_dir where model results will be saved
```
Comment (jds485) [inline 03b_model/src/run_model.py:299]: What happens if you use the same run_id for 2 runs? Will the 1st be overwritten, or will the code stop when it detects data in these directories?

File: 03b_model/src/run_model.py [hunk 3]
```diff
@@ -214,27 +449,66 @@ def train_model(prepped_model_io_data_file, inputs, seq_len,
...
+        end of validation in YYYY-MM-DD format
+    test_start_date : str
+        start of testing in YYYY-MM-DD format
+    test_end_date : str
+        end of testing in YYYY-MM-DD format
+
+    Returns
+    -------
+    trainval_df : TYPE
+        DESCRIPTION.
```
Comment (jds485) [inline 03b_model/src/run_model.py:509]: noting that the type and description are not filled out

File: 03b_model/src/run_model.py [hunk 4]
```diff
@@ -291,34 +581,183 @@ def plot_save_predictions(trainval_df, out_dir, run_id):
     
     #save predictions
     trainval_df.to_csv(os.path.join(out_dir, run_id, 'ModelResults.csv'))
+    
+def run_replicates(n_reps, prepped_model_io_data_file):
+    '''
+    run model replicates to understand variability
```
Comment (jds485) [inline 03b_model/src/run_model.py:587]: Suggested add to description: and saves the results and error plots to sub-directories in run_id

File: 03b_model/src/run_model.py [hunk 5]
```diff
@@ -291,34 +581,183 @@ def plot_save_predictions(trainval_df, out_dir, run_id):
...
+                             hidden_units, recur_dropout, dropout, 
+                             n_epochs, learn_rate, out_dir, run_id,
+                             train_start_date, train_end_date,
+                             val_start_date, val_end_date,
+                             test_start_date, test_end_date)
+        
+        plot_save_predictions(predictions, out_dir, run_id)
+        
+
+def test_hyperparameters():
```
Comment (jds485) [inline 03b_model/src/run_model.py:623]: Noting missing function description. Would help to note what kind of tuning is done (looks like grid search based on input values?)

Is this function currently unused?

**Comment by galengorski** [comment]
> If it’s not already, I think your run instructions could be in a README. Here are some possible edits to consider:

Thanks I'll create a README with this info

> 
> * I already had the repo cloned with older packages in the environment. I used this to update the environment and remove old/outdated dependencies
>   : `conda env update --file environment.yaml –prune`
> * It looks like the `Snakefile_fetch_munge` run worked, but I do see the following error. Should I be on VPN to run this?
> 
> ```
> Using IDP Account default to access ADFS https://fs.doi.gov/
> Authenticating as <user>@usgs.gov ...
> error authenticating to IdP: unable to classify response from auth server
> ```

I get this error whether I am on VPN or not, would be good to investigate

> * Seems like `river_dl` is missing from the environment. Let me know how that should be added. ModuleNotFoundErrorin line 4 of `Snakefile_b_ml_model_baseline`:
>   No module named 'river_dl'

Good catch, you can run `git submodule add https://github.com/USGS-R/river-dl.git` from the `03b_model/src/` directory. In the future this can be done during repo cloning with the `--recurse-submodules` command, which I will put in the README file as well.

**Review by jds485 — CHANGES_REQUESTED** [review]
Your edits look good to me! I have some small comments to consider. Also:

- pretrain – should a different term be used? Maybe initialize? `initialize_model`
- Do you set a random seed somewhere for each run?
- In all functions that write files or plots, the “Returns” section of the function header could list those. I think you have those in the function description already, though.

File: 03b_model/src/run_model.py (hunk 5 above)
Comment (galengorski) [reply 03b_model/src/run_model.py:623]: I updated the docstring in the latest commit. The function is used but not within the `main()` call, I call it directly to run the hyperparameter search

File: 03b_model/src/run_model.py (hunk 2 above)
Comment (galengorski) [reply 03b_model/src/run_model.py:299]: It will be overwritten, this is something that I need to add back in. You can see that I added some code to do this lines 344-351, but it was producing odd behavior with snakemake so I commented it out. I will add it back in and try to troubleshoot the behavior

File: 03b_model/src/run_model.py (hunk 1 above)
Comment (galengorski) [reply 03b_model/src/run_model.py]: Agreed! I was waiting on the methods for calculating the salt front before fully integrating into the pipeline because I thought we might be reproducing it ourselves. I have changed it to a variable

File: 03b_model/src/run_model.py (hunk 2 above)
Comment (jds485) [reply 03b_model/src/run_model.py:299]: I tried this out and it errored on creation of the snakemake file. Switching to a different run_id works.
```
WorkflowError:
Error recording metadata for finished job ([WinError 183] Cannot create a file when that file already exists: '\\.snakemake\\metadata\\tmpwwcy1f56.MDNiX21v' -> '\\.snakemake\\metadata\\MDNiX21vZGVsL291dC9pbnB1dHMuemFycg=='). Please ensure write permissions for the directory C:\Users\jsmith\OneDrive - DOI\Documents\Projects\DRB-Esal\CodeReviews\Galen\drb-estuary-salinity-ml\.snakemake
```

**Review by jds485 — APPROVED** [review]
